service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/ActionCaseManagementOrderController.java

```diff
@@ -0,0 +1,125 @@
+import uk.gov.hmcts.reform.fpl.model.CaseData;
+import uk.gov.hmcts.reform.fpl.model.CaseManagementOrder;
+import uk.gov.hmcts.reform.fpl.model.common.DocmosisDocument;
+import uk.gov.hmcts.reform.fpl.service.CMODocmosisTemplateDataGenerationService;
+import uk.gov.hmcts.reform.fpl.service.CaseManagementOrderService;
+import uk.gov.hmcts.reform.fpl.service.DocmosisDocumentGeneratorService;
+import uk.gov.hmcts.reform.fpl.service.DraftCMOService;
+import uk.gov.hmcts.reform.fpl.service.UploadDocumentService;
+
+import java.io.IOException;
+import java.util.Map;
+
+import static uk.gov.hmcts.reform.fpl.model.common.DocumentReference.buildFromDocument;
+
+@Api
+@RestController
+@RequestMapping("/callback/action-cmo")
+public class ActionCaseManagementOrderController {
+    private final DraftCMOService draftCMOService;
+    private final CaseManagementOrderService caseManagementOrderService;
+    private final DocmosisDocumentGeneratorService docmosisDocumentGeneratorService;
+    private final UploadDocumentService uploadDocumentService;
+    private final ObjectMapper mapper;
+    private final CMODocmosisTemplateDataGenerationService templateDataGenerationService;
+
+    public ActionCaseManagementOrderController(DraftCMOService draftCMOService,
+                                               CaseManagementOrderService caseManagementOrderService,
+                                               DocmosisDocumentGeneratorService docmosisDocumentGeneratorService,
+                                               UploadDocumentService uploadDocumentService,
+                                               ObjectMapper mapper,
+                                               CMODocmosisTemplateDataGenerationService templateDataGenerationService) {
+        this.draftCMOService = draftCMOService;
+        this.caseManagementOrderService = caseManagementOrderService;
+        this.docmosisDocumentGeneratorService = docmosisDocumentGeneratorService;
+        this.uploadDocumentService = uploadDocumentService;
+        this.mapper = mapper;
+        this.templateDataGenerationService = templateDataGenerationService;
+    }
+
+    @PostMapping("/about-to-start")
+    public AboutToStartOrSubmitCallbackResponse handleAboutToStart(@RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        Map<String, Object> data = caseDetails.getData();
+        final CaseData caseData = mapper.convertValue(data, CaseData.class);
+
+        caseDetails.getData()
+            .putAll(caseManagementOrderService.extractMapFieldsFromCaseManagementOrder(caseData.getCmoToAction()));
+
+        draftCMOService.prepareCustomDirections(caseDetails, caseData.getCmoToAction());
+
+        return AboutToStartOrSubmitCallbackResponse.builder()
+            .data(caseDetails.getData())
+            .build();
+    }
+
+    @PostMapping("/mid-event")
+    public AboutToStartOrSubmitCallbackResponse handleMidEvent(
```

Comment:
(0) not to be changes right now, but I think we need to be more explicit with our mid-event callbacks.

We can have multiple mid-events per flow, can't we? I believe we should name circumstances - so it's more obvious when this one is called. But not for this PR.

Reply:
I guess this is an issue everywhere. But yes you're right, now we have multi page flows, we can fire many mid events from one event. I am not sure we do this anywhere yet, but definitely one to watch out for 👍